<h2>Ý tưởng</h2>

<blockquote>Hãy cho tôi biết bạn của bạn là ai, tôi sẽ cho bạn biết bạn là người như thế nào.</blockquote>
Đó là một câu danh ngôn mà có lẽ ai cũng biết. Và có lẽ, phần lớn đều đồng ý. Đây là ý tưởng đằng sau thuật toán KNN. Để dự đoán mẫu dữ liệu mới $x_{new}$ thuộc về lớp nào, ta dựa vào số $k$ dữ liệu gần nó nhất. Ví dụ $k = 3$, nghĩa là gần $x_{new}$ có 3 điểm dữ liệu. Giả sử trong đó có 2 điểm dữ liệu thuộc về lớp B và 1 điểm dữ liệu thuộc về lớp A. Như vậy, ta sẽ gán $x_{new}$ thuộc về lớp B do $x_{new}$ có 2 điểm dữ liệu thuộc lớp B gần nó nhất.
<h2>Thuật toán</h2>



Thuật toán xác định lớp cho mẫu mới $x_{new}$
<ul>
	<li>Tính khoảng cách giữa $x_{new}$ và tất cả các mẫu trong tập huấn luyện</li>
	<li>Chọn k mẫu gần nhất với $x_{new}$ trong tập huấn luyện</li>
	<li>Gán $x_{new}$ vào lớp có nhiều mẫu nhất trong số k mẫu láng giềng đó (hoặc $x_{new}$ nhận giá trị trung bình của k mẫu)</li>
</ul>
<h2>Đánh giá</h2>

<strong>Ưu điểm</strong>
<ul>
	<li>Dễ sử dụng và cài đặt</li>
	<li>Xử lý tốt với dữ liệu nhiễu (do dựa trên khoảng cách để quyết định phân lớp)</li>
</ul>
<strong>Khuyết điểm</strong>
<ul>
	<li>Cần lưu tất cả các mẫu để dự đoán</li>
	<li>Cần nhiều thời gian để xác định lớp cho một mẫu mới (cần tính và so sánh khoảng cách đến tất cả các mẫu huấn luyện)</li>
	<li>Phụ thuộc vào giá trị k do người dùng lựa chọn. Nếu k quá nhỏ, nhạy cảm với nhiễu. Nếu k quá lớn, vùng lân cận có thể chứa các điểm của lớp khác.</li>
	<li>Cần chuyển đổi kiểu dữ liệu với thuộc tính phi số (categorial)</li>
</ul>
<h2>Ví dụ minh hoạ</h2>

Ta sẽ sử dụng thuật toán KNN của scikit-learn để phân lớp dữ liệu với tập huấn luyện là Iris. Tập dữ liệu hoa Iris được Ronald Fisher giới thiệu vào năm 1936. Tập dữ liệu này được dùng cho các bài toán phân lớp. Dữ liệu gồm 50 mẫu thu thập từ ba loại hoa Iris (Iris setosa, Iris virginica và Iris versicolor). Bốn thuộc tính cho mỗi mẫu gồm chiều dài và chiều rộng của đài hoa và cánh hoa được tính theo đơn vị centimet.


In [1]:
# k-Nearest Neighbor
from sklearn import datasets
from sklearn import metrics
from sklearn.neighbors import KNeighborsClassifier

# load iris the datasets
dataset = datasets.load_iris()

# 6 dòng dữ liệu đầu tiên của tập dữ liệu Iris
dataset.data[0:6]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4]])

Tiếp theo, ta sẽ cài đặt chương trình minh hoạ cho thuật toán KNN.

In [2]:
# fit a k-nearest neighbor model to the data
model = KNeighborsClassifier()
model.fit(dataset.data, dataset.target)
print(model)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')


In [3]:
# make predictions
expected = dataset.target
predicted = model.predict(dataset.data)

# summarize the fit of the model
print(metrics.classification_report(expected, predicted))

             precision    recall  f1-score   support

          0       1.00      1.00      1.00        50
          1       0.96      0.94      0.95        50
          2       0.94      0.96      0.95        50

avg / total       0.97      0.97      0.97       150



In [4]:
print(metrics.confusion_matrix(expected, predicted))

[[50  0  0]
 [ 0 47  3]
 [ 0  2 48]]


Qua ví dụ trên, ta thấy thuật toán KNN ứng dụng khá tốt trên tập dữ liệu Iris. KNN cho độ chính xác (precision) trung bình là 97%, (recall) trung bình là 97%, và (f1-score) là 97%. Để cài đặt thuật toán, ta chỉ cần khai báo lớp KNeighborsClassifier và gọi hàm fit() ứng với mô hình phân lớp này.